File: guess.hs
import Control.Exception
import Control.Monad.State
import System.Random

newtype Debug = Debug
  { errCount :: Integer
  }

debugDefault =
  Debug
    { errCount = 0
    }

data Game = Game
  { answer :: Integer,
    done :: Bool,
    guesses :: Integer,
    high :: Integer
  }

gameDefault =
  Game
    { answer = undefined,
      done = False,
      guesses = 0,
      high = undefined
    }

askGuess :: Integer -> IO Integer
askGuess high = do
  putStr $ "Guess a number between 1 and " ++ show high ++ ": "
  text <- getLine
  evaluate $ read text

askGuessMulti :: Integer -> StateT Debug IO Integer
askGuessMulti high =
  -- Use explicit ErrorCall to avoid catching Ctrl+C
  lift (askGuess high) `catchStateT` \(ErrorCall err) -> do
    lift $ putStrLn "I didn't understand"
    -- Alternatively use a global IORef for errCount.
    debug <- get
    put $ debug {errCount = errCount debug + 1}
    askGuessMulti high

-- From https://www.schoolofhaskell.com/user/snoyberg/general-haskell/exceptions/exceptions-and-monad-transformers
catchStateT ::
  Exception e => StateT s IO a -> (e -> StateT s IO a) -> StateT s IO a
catchStateT action handler = do
  state <- get
  (result, state') <-
    liftIO $
      runStateT action state `catch` \err ->
        runStateT (handler err) state
  put state'
  return result

pickAnswer :: Integer -> IO Integer
pickAnswer high =
  randomRIO (1, high)

pickAnswer2 :: RandomGen gen => Integer -> gen -> (Integer, gen)
pickAnswer2 high = uniformR (1, high)

play :: Game -> StateT Debug IO Game
play game = do
  guess <- askGuessMulti $ high game
  lift $ game `report` guess
  let next = game `update` guess
  if done next then return next else play next

report :: Game -> Integer -> IO ()
report game guess =
  putStrLn $ show guess ++ " is " ++ description
  where
    description = case compare guess (answer game) of
      LT -> "too low"
      GT -> "too high"
      EQ -> "the answer!"

update :: Game -> Integer -> Game
update game guess =
  game {done = guess == answer game, guesses = guesses game + 1}

main :: IO ()
main = do
  answer <- pickAnswer high
  -- let (answer, gen2) = pickAnswer2 high $ mkStdGen 1337
  -- answer <- getStdRandom $ pickAnswer2 high
  let game = gameDefault {answer = answer, high = high}
  (result, debug) <- runStateT (play game) debugDefault
  putStrLn $ "Finished in " ++ show (guesses result) ++ " guesses"
  putStrLn $ "Total input errors: " ++ show (errCount debug)
  where
    high = 100

File: Guess.java
import java.util.Random;
import static java.lang.System.*;

class Guess {
    static int pickAnswer(Random random, int high) {
        return random.nextInt(high) + 1;
    }

    public static void main(String[] args) {
        var high = 100;
        var answer = pickAnswer(new Random(), high);
        var game = new Game(answer, high);
        game.play();
        out.printf("Finished in %s guesses\n", game.guesses());
        err.printf("Total input errors: %s\n", game.errCount());
    }
}

class Game {
    private int answer;
    private boolean done;
    private int errCount;
    private int guesses;
    private int high;

    public Game(int answer, int high) {
        this.answer = answer;
        this.high = high;
    }

    int askGuess() throws NumberFormatException {
        out.printf("Guess a number between 1 and %s: ", high);
        return Integer.parseInt(console().readLine());
    }

    int askGuessMulti() {
        while (true) {
            try {
                return askGuess();
            } catch (NumberFormatException e) {
                out.println("I didn't understand");
                errCount += 1;
            }
        }
    }

    int errCount() {
        return errCount;
    }

    public int guesses() {
        return guesses;
    }

    public void play() {
        while (!done) {
            var guess = askGuessMulti();
            report(guess);
            update(guess);
        }
    }

    void report(int guess) {
        var description =
            guess < answer ? "too low" :
            guess > answer ? "too high" :
            "the answer!";
        out.printf("%s is %s\n", guess, description);
    }

    void update(int guess) {
        if (guess == answer) {
            done = true;
        }
        guesses += 1;
    }
}

// record Game(
//     int answer,
//     boolean done,
//     int guesses,
//     int high
// ) {
//     public Game(int answer, int high) {
//         this(answer, false, 0, high);
//     }
// }
